src/lang/diagnostics/mod.rs

```diff
@@ -109,29 +114,41 @@ impl DiagnosticsControllerThread {
     /// Runs a single tick of the diagnostics controller's event loop.
     #[tracing::instrument(skip_all)]
     fn diagnostics_controller_tick(&self, state_snapshots: StateSnapshots) {
+        let jobs_number = state_snapshots.0.len();
+        let (worker_status_sender, worker_status_receiver) =
+            crossbeam::channel::bounded(jobs_number);
         let (state, primary_snapshots, secondary_snapshots) = state_snapshots.split();
 
         let primary_set = find_primary_files(&state.db, &state.open_files);
         let primary: Vec<_> = primary_set.iter().copied().collect();
-        self.spawn_refresh_worker(&primary, primary_snapshots);
+        self.spawn_refresh_worker(&primary, primary_snapshots, worker_status_sender.clone());
 
         let secondary = find_secondary_files(&state.db, &primary_set);
-        self.spawn_refresh_worker(&secondary, secondary_snapshots);
+        self.spawn_refresh_worker(&secondary, secondary_snapshots, worker_status_sender.clone());
 
         let files_to_preserve: HashSet<Url> = primary
             .into_iter()
             .chain(secondary)
             .flat_map(|file| state.db.url_for_file(file))
             .collect();
 
-        self.spawn_worker(move |project_diagnostics, notifier| {
-            clear_old_diagnostics(files_to_preserve, project_diagnostics, notifier);
-        });
+        self.spawn_worker(
+            move |project_diagnostics, notifier| {
+                clear_old_diagnostics(files_to_preserve, project_diagnostics, notifier);
+            },
+            worker_status_sender.clone(),
+        );
+
+        worker_status_receiver.iter().nth(jobs_number - 1);
     }
 
     /// Shortcut for spawning a worker task which does the boilerplate around cloning state parts
     /// and catching panics.
-    fn spawn_worker(&self, f: impl FnOnce(ProjectDiagnostics, Notifier) + Send + 'static) {
+    fn spawn_worker(
+        &self,
+        f: impl FnOnce(ProjectDiagnostics, Notifier) + Send + 'static,
+        sender: crossbeam::channel::Sender<WorkerStatus>,
```

Comment:
```suggestion
        sender: crossbeam::channel::Sender<()>,
```

TBH I don't see a need to declare a custom ZST for this. I would rather focus on encapsulating entire waiting logic into some kind of `TaskHandle` with a `join` method, like we have for threads

Reply:
Done

Reply:
I believe @Draggu did it this way in the first diagnostic impls. It was for better reasoning about what to send if sth was cancelled etc. (e.g. we can still send sth if cancellation happened if diags for some files were calculated before cancellation etc.)

Reply:
In first impl we were sending update and then applying it to state that lived on diagnostics main thread, now we are using `RwLock` for this so it does not matter

Reply:
@piotmag769 @mkaput can we resolve?

Reply:
yes, we can